File: src/card/markdown-render.ts
import { card, hr, image, md, note, type CardElement, type CardObject, type HeaderTemplate } from './cards';

/**
 * Markdown → card-element rendering for outbound replies. Two jobs:
 *
 *  1. {@link renderRichText} — turn a markdown answer into card elements,
 *     splitting out `![alt](src)` images (resolved to `img_key` via the upload
 *     map) into real `img` elements interleaved with the text. Unresolved
 *     sources keep their literal markdown (graceful degradation). It also
 *     strips ```feishu-card fences, which are hoisted into their own card.
 *
 *  2. {@link buildCleanCard} — parse one ```feishu-card fence's markdown into a
 *     standalone card: a leading heading becomes the card header, `---` → hr,
 *     `> …` → grey note, everything else → markdown (with inline images). The
 *     bridge owns this mapping so the emitted card is always valid schema 2.0 —
 *     codex only writes markdown, never hand-rolled (often wrong) card JSON.
 *
 * Both share the same image-aware text splitter and the `src → image_key` map
 * produced by {@link ./outbound-images}.
 */

type ImageMap = ReadonlyMap<string, string>;
const NO_IMAGES: ImageMap = new Map();

/** `![alt](src)` — group 1 = alt, group 2 = src (possibly `<…>`-wrapped). Mirrors
 * the source-scanning regex in outbound-images so the two stay in lockstep. */
const IMG_RE = /!\[([^\]]*)\]\(\s*(<[^>]+>|[^)\s]+)(?:\s+(?:"[^"]*"|'[^']*'))?\s*\)/g;
/** A ```feishu-card fenced block; group 1 = the inner markdown. */
const FENCE_RE = /```feishu-card[^\n]*\n([\s\S]*?)```/g;

function cleanSrc(raw: string): string {
  let s = raw.trim();
  if (s.startsWith('<') && s.endsWith('>')) s = s.slice(1, -1).trim();
  return s;
}

/**
 * Pull every ```feishu-card fence out of `text`. Returns the fences' inner
 * markdown (trimmed) and `text` with the fences removed (so the run card never
 * shows a card spec as a raw code block — it's rendered as a clean card
 * instead).
 */
export function extractCardFences(text: string): { fences: string[]; stripped: string } {
  const fences: string[] = [];
  const re = new RegExp(FENCE_RE.source, 'g');
  const stripped = text.replace(re, (_full, inner: string) => {
    fences.push(inner.trim());
    return '';
  });
  return { fences, stripped };
}

/**
 * Render a markdown string into card elements, replacing resolved `![](src)`
 * images with `img` elements and stripping any ```feishu-card fences. Plain
 * text (no images, no fences) short-circuits to a single markdown element.
 */
export function renderRichText(text: string, images: ImageMap = NO_IMAGES): CardElement[] {
  const body = extractCardFences(text).stripped;
  if (!body.includes('![')) {
    const t = body.trim();
    return t ? [md(t)] : [];
  }
  const els: CardElement[] = [];
  let buf = '';
  const flush = (): void => {
    const t = buf.trim();
    if (t) els.push(md(t));
    buf = '';
  };
  const re = new RegExp(IMG_RE.source, 'g');
  let last = 0;
  let m: RegExpExecArray | null;
  while ((m = re.exec(body)) !== null) {
    buf += body.slice(last, m.index);
    const alt = m[1] ?? '';
    const src = cleanSrc(m[2] ?? '');
    const key = images.get(src);
    if (key) {
      flush();
      els.push(image(key, alt));
    } else {
      // Unresolved (path outside cwd / fetch failed / not yet uploaded): keep
      // the literal markdown so the reference isn't silently dropped.
      buf += m[0];
    }
    last = m.index + m[0].length;
  }
  buf += body.slice(last);
  flush();
  return els;
}

/**
 * Build a standalone clean card from one ```feishu-card fence's markdown. A
 * leading heading (`# …`) becomes the card header (blue); the rest is split
 * into blocks on blank lines and mapped element-by-element.
 */
export function buildCleanCard(
  fenceMarkdown: string,
  images: ImageMap = NO_IMAGES,
  template: HeaderTemplate = 'blue',
): CardObject {
  const lines = fenceMarkdown.split('\n');
  let start = 0;
  while (start < lines.length && lines[start]?.trim() === '') start++;
  const headingMatch = lines[start]?.match(/^#{1,6}\s+(.+?)\s*$/);
  const title = headingMatch ? headingMatch[1] : '';
  if (headingMatch) start++;

  const bodyMarkdown = lines.slice(start).join('\n').trim();
  const elements = renderCleanBody(bodyMarkdown, images);
  // A card needs at least one body element — fall back to the title (or a
  // spacer) so a title-only fence still produces a valid card.
  const body = elements.length > 0 ? elements : [md(title || '­')];

  return card(body, {
    ...(title ? { header: { title, template } } : {}),
    summary: title || '卡片',
  });
}

/** Split clean-card body markdown into blocks (on blank lines) and map each to
 * an element: a markdown rule → hr, a `> …` quote block → note, anything else →
 * image-aware markdown. */
function renderCleanBody(bodyMarkdown: string, images: ImageMap): CardElement[] {
  const out: CardElement[] = [];
  for (const raw of bodyMarkdown.split(/\n{2,}/)) {
    const block = raw.trim();
    if (!block) continue;
    if (/^(-{3,}|\*{3,}|_{3,})$/.test(block)) {
      out.push(hr());
      continue;
    }
    const blockLines = block.split('\n');
    if (blockLines.every((l) => l.trim() === '' || /^\s*>\s?/.test(l))) {
      const noteText = blockLines
        .map((l) => l.replace(/^\s*>\s?/, ''))
        .join('\n')
        .trim();
      if (noteText) out.push(note(noteText));
      continue;
    }
    out.push(...renderRichText(block, images));
  }
  return out;
}

File: internal/card/markdown_render.go
package card

// markdown_render.go —— markdown → 卡片元素渲染（对齐 TS card/markdown-render）。
// 两个职责：renderRichText（markdown+图片→元素）、buildCleanCard（feishu-card 围栏→独立卡）。
// 纯函数，依赖 element builder + outbound imgRe，零飞书 SDK 调用。

import (
	"regexp"
	"strings"
)

var (
	fenceRe     = regexp.MustCompile("```feishu-card[^\\n]*\\n([\\s\\S]*?)```")
	headingRe   = regexp.MustCompile(`^#{1,6}\s+(.+?)\s*$`)
	hrBlockRe   = regexp.MustCompile(`^(-{3,}|\*{3,}|_{3,})$`)
	quoteLineRe = regexp.MustCompile(`^\s*>\s?`)
)

// FencesResult extractCardFences 结果。
type FencesResult struct {
	Fences   []string // 围栏内 markdown（trim）
	Stripped string   // 移除围栏后的文本
}

// ExtractCardFences 抽出全部 ```feishu-card 围栏（返回内部 markdown + 移除围栏的文本）。
func ExtractCardFences(text string) FencesResult {
	var fences []string
	stripped := fenceRe.ReplaceAllStringFunc(text, func(full string) string {
		m := fenceRe.FindStringSubmatch(full)
		if len(m) >= 2 {
			fences = append(fences, strings.TrimSpace(m[1]))
		}
		return ""
	})
	return FencesResult{Fences: fences, Stripped: stripped}
}

// RenderRichText markdown → 元素：已上传图片换 Image 元素，未解析保留原文，剥 feishu-card 围栏。
// 纯文本（无 ![）短路为单个 Md 元素。
func RenderRichText(text string, images map[string]string) []CardElement {
	body := ExtractCardFences(text).Stripped
	if !strings.Contains(body, "![") {
		t := strings.TrimSpace(body)
		if t == "" {
			return nil
		}
		return []CardElement{Md(t)}
	}
	var els []CardElement
	buf := ""
	flush := func() {
		if t := strings.TrimSpace(buf); t != "" {
			els = append(els, Md(t))
		}
		buf = ""
	}
	locs := imgRe.FindAllStringSubmatchIndex(body, -1)
	last := 0
	for _, loc := range locs {
		fullStart, fullEnd := loc[0], loc[1]
		altStart, altEnd, srcStart, srcEnd := loc[2], loc[3], loc[4], loc[5]
		buf += body[last:fullStart]
		alt := body[altStart:altEnd]
		src := CleanSrc(body[srcStart:srcEnd])
		if key, ok := images[src]; ok {
			flush()
			els = append(els, Image(key, alt))
		} else {
			// 未解析（cwd 外 / 上传失败）：保留原文，绝不静默丢引用。
			buf += body[fullStart:fullEnd]
		}
		last = fullEnd
	}
	buf += body[last:]
	flush()
	return els
}

// BuildCleanCard 从一个 feishu-card 围栏的 markdown 构造独立卡：
// 首行 # 标题 → header（蓝），其余按空行分块映射元素（--- → hr、> 引用 → note、其它 → 图片感知 markdown）。
func BuildCleanCard(fenceMarkdown string, images map[string]string, template HeaderTemplate) CardObject {
	lines := strings.Split(fenceMarkdown, "\n")
	start := 0
	for start < len(lines) && strings.TrimSpace(lines[start]) == "" {
		start++
	}
	title := ""
	if start < len(lines) {
		if m := headingRe.FindStringSubmatch(lines[start]); len(m) >= 2 {
			title = strings.TrimSpace(m[1])
			start++
		}
	}
	bodyMd := strings.TrimSpace(strings.Join(lines[start:], "\n"))
	elements := renderCleanBody(bodyMd, images)
	if len(elements) == 0 {
		fallback := title
		if fallback == "" {
			fallback = ""
		}
		elements = []CardElement{Md(fallback)}
	}
	summary := title
	if summary == "" {
		summary = "卡片"
	}
	opts := CardOpts{Summary: summary}
	if title != "" {
		opts.Header = &CardHeader{Title: title, Template: template}
	}
	return Card(elements, opts)
}

// renderCleanBody 按空行分块映射：分隔线 → hr；引用块 → note；其余 → renderRichText。
func renderCleanBody(bodyMarkdown string, images map[string]string) []CardElement {
	var out []CardElement
	for _, raw := range regexp.MustCompile(`\n{2,}`).Split(bodyMarkdown, -1) {
		block := strings.TrimSpace(raw)
		if block == "" {
			continue
		}
		if hrBlockRe.MatchString(block) {
			out = append(out, Hr())
			continue
		}
		blockLines := strings.Split(block, "\n")
		allQuote := true
		for _, l := range blockLines {
			if strings.TrimSpace(l) != "" && !quoteLineRe.MatchString(l) {
				allQuote = false
				break
			}
		}
		if allQuote {
			stripped := []string{}
			for _, l := range blockLines {
				stripped = append(stripped, quoteLineRe.ReplaceAllString(l, ""))
			}
			noteText := strings.TrimSpace(strings.Join(stripped, "\n"))
			if noteText != "" {
				out = append(out, Note(noteText))
			}
			continue
		}
		out = append(out, RenderRichText(block, images)...)
	}
	return out
}
